jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketUpgradeHandler.java

```diff
@@ -120,9 +121,10 @@ public boolean handle(Request request, Response response, Callback callback) thr
     @Override
     public InvocationType getInvocationType()
     {
-        // Must be BLOCKING because EndPoint.upgrade() ends up invoking
-        // application code in the WebSocket "connect" event handler,
-        // e.g. a method annotated with @OnWebSocketConnect, that may block.
-        return InvocationType.BLOCKING;
+        if (isDynamic())
+            return InvocationType.BLOCKING;
+        Handler handler = getHandler();
+        InvocationType result = handler == null ? InvocationType.NON_BLOCKING : handler.getInvocationType();
+        return Invocable.combine(result, container.getInvocationType());
```

Comment:
Tiny quibble with the name. Maybe better to have:
```suggestion
        InvocationType handlerInvocationType = handler == null ? InvocationType.NON_BLOCKING : handler.getInvocationType();
        return Invocable.combine(handlerInvocationType, container.getInvocationType());
```